src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs

```diff
@@ -89,54 +89,84 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
+                // Set the shell state to 'Initializing' to guarantee that any subsequent HTTP requests
+                // will result in a 'Service Unavailable' response while the tenant setup is in progress.
+                context.ShellSettings.AsInitializing();
 
-                if (context.Errors.Any())
+                (var executionId, shellSettings) = await SetupInternalAsync(context);
+
+                if (context.Errors.Count > 0)
                 {
                     context.ShellSettings.State = initialState;
-                    await _shellHost.ReloadShellContextAsync(context.ShellSettings, eventSource: false);
+                    await _shellHost.ReloadShellContextAsync(shellSettings, eventSource: false);
+
+                    var shellScope = await _shellHost.GetScopeAsync(shellSettings);
+                    await shellScope.UsingAsync(async scope =>
+                    {
+                        var tenandSetupHandlers = scope.ServiceProvider.GetServices<ITenantSetupHandler>();
+
+                        await tenandSetupHandlers.InvokeAsync((handler) => handler.FailedAsync(context), _logger);
+                    });
+
+                    return null;
                 }
 
+                // At this point, we know that the setup was successful.
+                // Set the shell state to 'Running' state.
+                await _shellHost.UpdateShellSettingsAsync(shellSettings.AsRunning());
+
+                var runningShellScope = await _shellHost.GetScopeAsync(shellSettings);
+                await runningShellScope.UsingAsync(async scope =>
+                {
+                    var tenandSetupHandlers = scope.ServiceProvider.GetServices<ITenantSetupHandler>();
+
+                    await tenandSetupHandlers.InvokeAsync((handler) => handler.SucceededAsync(), _logger);
+                });
+
                 return executionId;
             }
-            catch
+            catch (Exception e)
             {
+                _logger.LogError(e, "An error occurred while setting up the tenant.");
+                context.Errors.Add(String.Empty, S["An error occurred while setting up the tenants: {0}", e.Message]);
+
                 context.ShellSettings.State = initialState;
-                await _shellHost.ReloadShellContextAsync(context.ShellSettings, eventSource: false);
+                await _shellHost.ReloadShellContextAsync(shellSettings ?? context.ShellSettings, eventSource: false);
+
+                var shellScope = await _shellHost.GetScopeAsync(shellSettings ?? context.ShellSettings);
+                await shellScope.UsingAsync(async scope =>
+                {
+                    var tenandSetupHandlers = scope.ServiceProvider.GetServices<ITenantSetupHandler>();
+
+                    await tenandSetupHandlers.InvokeAsync((handler) => handler.FailedAsync(context), _logger);
+                });
 
-                throw;
+                return null;
```

Comment:
Not sure I agree, in this catch block I would still only reload the shell and then re-throw.

Reply:
Are you sure about this? I am looking at how we call `SetupAsync() from `AutoSetupMiddleware` or `SetupController` and we are not catching that error if an exception was thrown. With the change, we log the error into the `Errors` collection and we prevent a system exception.

Reply:
@jtkech after reading my last comment, do you still think we should `throw` or `return executionId` is enough?